test/integration/scheduler_perf/scheduler_perf.go

```diff
@@ -641,6 +641,26 @@ func initTestOutput(tb testing.TB) io.Writer {
 	return output
 }
 
+type cleanupKeyType struct{}
+
+var cleanupKey = cleanupKeyType{}
+
+// shouldCleanup returns true if a function should clean up resource in the
+// apiserver when the test is done. This is true for unit tests (etcd and
+// apiserver get reused) and false for benchmarks (each benchmark starts with a
+// clean slate, so cleaning up just costs time).
+func getCleanup(ctx context.Context) bool {
+	val := ctx.Value(cleanupKey)
+	if enabled, ok := val.(bool); ok {
+		return enabled
+	}
+	return false
+}
+
+func withCleanup(tCtx ktesting.TContext, enabled bool) ktesting.TContext {
```

Comment:
This wrapper seems hard to maintain as independent.
Can this be an initOption, then we can judge here https://github.com/kubernetes/kubernetes/blob/8c80c07e85833c5b426fdd774d2729733d14adb6/test/utils/ktesting/tcontext.go#L375-L378

Reply:
No, I don't think this belongs into `ktesting` itself. Right now there's only a single user of it (scheduler_perf) and there it can be done via the context value mechanism.

Why is this hard to maintain? It's using a stable API of `ktesting` and `context`, so there's no risk that this is going to break.

Reply:
>  I don't think this belongs into ktesting itself.

But this decides whether we should call cleanupCtx() at the last, make it part of the ktesting seems easier to use, or we have to call the withCleanup() each time, looks a little weird. And can this be True by default, since cleanup is mostly right.

```
tCtx := ktesting.WithTB(tCtx, t)
tCtx = withCleanup(tCtx, true)
```

Agree with you that `Right now there's only a single user of it (scheduler_perf) and there it can be done via the context value mechanism.`, so leave this for further improvement(if needed) also makes sense to me. We can't make it perfect at one time.

Reply:
> make it part of the ktesting seems easier to use, we have to call the withCleanup() each time

Hmm, I'm not sure I follow. Whether it's part of ktesting or layered on top of it doesn't change the numbers of calls to such a `withCleanup` or `ktesting.WithCleanup` method.

> And can this be True by default, since cleanup is mostly right.

Okay.

Reply:
I mean we have to call `tCtx = withCleanup(tCtx, true)` each time additionally. And add judgement here each time.
```
if err == nil && getCleanup(tCtx) {
}
```

Reply:
But yes we have only one test now.

Reply:
So i'm ok to leave as it is right now until we have more tests. But that's my poor points.

Reply:
> I mean we have to call tCtx = withCleanup(tCtx, true) each time additionally. And add judgement here each time.

So you wanted to make the behavior of `tCtx.CleanupCtx` itself depend on this context setting? We cannot do that because some other cleanup callbacks still need to be invoked. It's really only the deletion of Kubernetes API resources which may get skipped, and that is not something that can be detected inside `ktesting`.